pandas/core/indexes/base.py

```diff
@@ -2654,12 +2658,16 @@ def _union(self, other, sort):
                         stacklevel=3,
                     )
 
-        # for subclasses
-        return self._wrap_setop_result(other, result)
+        return self._shallow_copy(result)
 
     def _wrap_setop_result(self, other, result):
         name = get_op_result_name(self, other)
-        return self._shallow_copy(result, name=name)
+        if isinstance(result, ABCIndexClass):
```

Comment:
this check will include Index objects but exclude subclasses.  is that intentional?  if so, can you add a comment about why

Reply:
I'll admit I don't actually understand this whole ABC business 100%, but I'm getting:
```
isinstance(pd.MultiIndex, pd.Index) -> False
isinstance(pd.MultiIndex, ABCIndexClass) -> True
issubclass(pd.MultiIndex, pd.Index) -> True
```
Should I be using `issubclass` instead? Wasn't trying to do anything too fancy here.

Reply:
yah its not intuitive.  basically `isinstance(obj, ABCIndexClass)` is equivalent to `type(obj) is Index`, so what you probably want is `isinstance(obj, Index)`, which behaves like normal

Reply:
Hmm, but wouldn't we get `isinstance(multiindex, ABCIndexClass) == True` and `type(multiindex) != pd.Index` here? Either way, agreed that `isinstance` will get us what we want.

Reply:
my bad, i confused ABCIndexClass with ABCIndex, never mind.